Deployment docs (Cloudstate managed services & self-hosted) (#2)
cloudstateio/samples-ui-chat: README.adoc

= Cloudstate Sample Chat Application
Lightbend, 2020
Version 2.0, 2020-06-25
:description: Cloudstate Chat Application 
:keywords: Cloudstate, stateful, serverless, chat-sample, tutorial
:sectnums:
:toc:
ifdef::env-github[]
:tip-caption: :bulb:
:note-caption: :information_source:
:important-caption: :heavy_exclamation_mark:
:caution-caption: :fire:
:warning-caption: :warning:
endif::[]

[[cs-chat-sample-overview]]
== Overview

> This is a sample application showing an example of using CloudState to build a chat application in Node.js. The application consists of three services:

Author: Done, thx

deploy:: directory containing the yaml files to deploy all the services composing the Chat application. The target cluster could be your own Kubernetes cluster or https://docs.lbcs.dev/index.html[Lightbend Cloudstate managed service]

The code of the backend services `Friends` and `Presence` is separate from the UI code. This is intentional to demonstrate the polyglot aspect of Cloudstate services. They can seamlessly interact regardless of the programming language used to implement the service. Below are the available implementations. You can choose any combination for deployment. For example, `Friends` service in Javascript and `Presence` service in Java.

=== Javascript implementation

* https://github.com/cloudstateio/samples-js-chat/tree/master/friends[Friends service]
* https://github.com/cloudstateio/samples-js-chat/tree/master/presence[Presence service]

=== Java implementation

* https://github.com/cloudstateio/samples-java-chat/tree/master/friends[Friends service]
* https://github.com/cloudstateio/samples-java-chat/tree/master/presence[Presence service]

== Prerequisites

> === Javascript Dev environment

Author: Changed to JavaScript (capital S)

https://github.com/cloudstateio/samples-js-chat/blob/master/README.adoc#js-devenv-setup-for-cloudstate[Javascript Dev environment setup for Cloudstate]

=== Infrastructure

* A Docker repository that the Kubernetes cluster has the necessary credentials to pull from. You can get a free Docker registry at https://hub.docker.com/

For the deployment, there are two possible choices. Below we highlight the main characteristic of each deployment approach. Further details about prerequisites and infrastructure preparation would be discussed further in the <<deployment,Deployment>> section. 

==== Deploy on Lighbend Cloudstate managed service

Basically, all you need is to open an account with Lightbend Cloudstate hosting service. Then install the "Cloudstate CLI" tool and you are good to deploy. There is no need to provide any local infrastructure.

> The deployment is as simple as applying the yaml configuration files. Lightbend Cloudstate managed service will take care of the hosting, configuration and maintenance of your service. 

Reviewer: I think there are no yaml files any longer, everything is done through the UI or the csctl utility. Maybe you could just link to the instructions from here.

Author: Oh good point. Thx. WIll update accordingly

=== Building the Friends & Presence services
* https://github.com/cloudstateio/samples-js-chat/blob/master/friends/README.adoc#building-the-friends-service[Building the Friends service]

* https://github.com/cloudstateio/samples-js-chat/blob/master/presence/README.adoc#building-the-presence-service[Building the Presence service]

=== Building the Chat-UI service

The code below assumes that your docker registry is named `mydockerregistry`. Please replace it with your DockerID (which you have registed at https://hub.docker.com/ as mentioned in <<infrastructure>>

[source,shell]
----
cd ./chat-ui

nvm install
nvm use
npm install
./protogen.sh
npm run prestart
npm run-script build

# let's assume your DockerID in https://hub.docker.com/ is `mydockerregistry`
DOCKER_PUBLISH_TO=mydockerregistry

# build docker image
docker build . -t $DOCKER_PUBLISH_TO/samples-js-chat-ui:latest

# authenticate with your Docker registry
docker login

# push the docker image to your registry
docker push $DOCKER_PUBLISH_TO/samples-js-chat-ui:latest
----

=== Testing the Chat-UI service

> In order to make the deployment of the Chat UI consistent with all the other services composing the Chat application. The Node.js server running the UI code is wrapped in a Cloudstate stateless service.

Reviewer 2: With "In order to make" something has to follow then, yes?

> In order to make the deployment of the Chat UI consistent with all the other services composing the Chat application. The Node.js server running the UI code is wrapped in a Cloudstate stateless service.

Reviewer 2: "Chat application" is named inconsistently within phrases, sometimes with "**C**hat application" and then as "**c**hat application"

As such, the UI must be accessed via a Cloudstate proxy. The procedure is similar to that of https://github.com/cloudstateio/samples-js-chat/blob/master/friends/README.adoc#testing-friends-service[Testing Friends service]. Which means:

> * Start a Cloudstate proxy (docker image prebuilt for us by Cloudstate dev team)

Reviewer 2: ```suggestion
* Start a Cloudstate proxy (docker image prebuilt for us by the Cloudstate project)
```

* Start the docker image of the Chat UI we have just built
* Access the UI via the Cloudstate proxy, which expose the Chat-UI service on `http://localhost:9000/pages/chat.html`

[source,shell]
----
# terminal 1: run the cloudstate proxy docker image
# exposing the CS proxy port 9000 to the same 9000 port on the host machine
# NOTE: ignore the connection errors b/c the `user-function` is not yet started
docker run -it --rm --name cloudstate-proxy -p 9000:9000 \

>   cloudstateio/cloudstate-proxy-dev-mode \

Reviewer 2: `cloudstateio/cloudstate-proxy-dev-mode:latest` is the build from master, no?

Author: I don't know well this part, if there is a recommended Cloudstate proxy other than `cloudstateio/cloudstate-proxy-dev-mode:latest` can you please suggest?

Reviewer 3: Yes, `latest` is the build from master. It could be better to have `latest` point to the latest release instead, and use a different tag for the builds from master. For now, how about using the latest release: `cloudstateio/cloudstate-proxy-dev-mode:0.5.1`

Author: By changing to `cloudstateio/cloudstate-proxy-dev-mode:0.5.1`, the documentation may get out-of-date in a near future upon a new release of the Cloudstate proxy. Can you confirm which approach is more preferable?
1. `latest`: doc always up-to-date but may run into instabilities b/c Cloudstate proxy build on `master` is not of release quality
2. `0.5.1` is guarantee stable, but we must remember to update all docs. The difficulty here is that the maintainers of the samples code are likely not the same as those maintaining Cloudstate.

Reviewer 3: The sample depends on a particular version of cloudstate node-support (currently `0.0.3`). If changes are made to the proxy, then the sample may not work. So I think it's better to have a stable version for the proxy which is known to work. Since this is described as a maintenance issue, we should focus there — ideally we have automated or easy ways to update all the samples after releases.

[[prereq-docker-registry]]
=== Docker repository

You must package your services into docker images and publish these images in a Docker registry. At deployment time, the Kubernetes cluster will pull the images from this registry. You can get a free Docker registry at https://hub.docker.com/

=== Self-hosting or Managed hosting?

For the deployment, there are two possible choices. Below we highlight the main characteristics of each deployment approach. Details about prerequisites and infrastructure preparation would be discussed further in the <<deployment,Deployment>> section. 

==== Managed hosting

This deployment approach doesn't require any local infrastructure. Lightbend Cloudstate will take care of the hosting, configuration and maintenance of your services. The deployment prodedure is pretty minimal:

* Open an account with Lightbend Cloudstate
* Install the "Cloudstate CLI" tool `csctl`

And that's it. Then you just use `csctl` to deploy the your services that you had published in your Docker registry.

 
==== Self-hosting

> In case you prefer to host your own services. You need to provide a Kubernetes cluster. This cluster could be local or cloud-based and must fulfill the following requirements:

Reviewer 4: ```suggestion
In case you prefer to host your own services, you need to provide a Kubernetes cluster. This cluster could be local or cloud-based and must fulfill the following requirements:
```

* A Kubernetes cluster, with 1.13 as the minimum required version
* An ingress controller with a Layer7 Load Balancer ability
* A working knowledge of Kubernetes to be able to manage your cluster and to troubleshoot, should any issue specific to your Kubernetes cluster arise.

== Build Docker images

Before beginning the deployment, each of the various microservices composing the Chat application must be packaged into their own docker image, and published to your Docker registry.

=== Building the Friends & Presence services
* https://github.com/cloudstateio/samples-js-chat/blob/master/friends/README.adoc#building-the-friends-service[Building the Friends service]

* https://github.com/cloudstateio/samples-js-chat/blob/master/presence/README.adoc#building-the-presence-service[Building the Presence service]

=== Building the Chat-UI service

> The code below assumes that your Docker registry is named `mydockerregistry`. Please replace it with your own DockerID (which you have registed at https://hub.docker.com/ as mentioned in <<prereq-docker-registry,Prerequisites - Docker repository>>

Reviewer 4: ```suggestion
The code below assumes that your Docker registry is named `mydockerregistry`. Please replace it with your own DockerID (which you might have registed at https://hub.docker.com/ as mentioned in <<prereq-docker-registry,Prerequisites - Docker repository>>
```
